### ISSUE TITLE
Miner process should recover from DuplicateKernel

### ISSUE THREAD

**Issue by ignopeverell** [mimblewimble/grin#507]
Guessing this has something to do with a key being reused by the wallet. But regardless it shouldn't crash the whole thing:

```
Dec 17 14:39:26.044 WARN P2P server started on 0.0.0.0:13414
Dec 17 14:39:26.049 WARN Grin server started.
thread 'miner' panicked at 'Box<Any>', grin/src/miner.rs:599:4
Dec 17 14:39:56.754 stack backtrace:
ERRO Error setting sumtree root to build a block: DuplicateKernel(Commitment(095a861643794d1c854f064da014a2760380bbfe16a69a93b369c978fdb0ea4095))
   0:     0x557ec62f8503 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::hfc7985b08e763a82
                               at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x557ec62f44f4 - std::sys_common::backtrace::_print::h16a1db02a59ead63
                               at /checkout/src/libstd/sys_common/backtrace.rs:71
   2:     0x557ec62fba73 - std::panicking::default_hook::{{closure}}::h48ecee46f2eefc30
                               at /checkout/src/libstd/sys_common/backtrace.rs:60
                               at /checkout/src/libstd/panicking.rs:381
   3:     0x557ec62fb7e2 - std::panicking::default_hook::hb4c92ae8d005ca44
                               at /checkout/src/libstd/panicking.rs:397
   4:     0x557ec62fbf37 - std::panicking::rust_panic_with_hook::h25d461655d60b1a5
                               at /checkout/src/libstd/panicking.rs:611
   5:     0x557ec4c94449 - std::panicking::begin_panic::hc7980fb899a9ae73
                               at /checkout/src/libstd/panicking.rs:572
   6:     0x557ec4e568ba - grin_grin::miner::Miner::build_block::h50ccdac7ded23bfd
                               at grin/src/miner.rs:599
   7:     0x557ec4e53a03 - grin_grin::miner::Miner::run_loop::h2e1d64d84a980999
                               at grin/src/miner.rs:450
   8:     0x557ec4e41df6 - grin_grin::server::Server::start_miner::{{closure}}::hb649cc1486ca9c15
                               at grin/src/server.rs:226
   9:     0x557ec4c7837a - std::sys_common::backtrace::__rust_begin_short_backtrace::hb991bd8e33204edc
                               at /checkout/src/libstd/sys_common/backtrace.rs:136
  10:     0x557ec4c93ff3 - std::thread::Builder::spawn::{{closure}}::{{closure}}::ha855a6e5c2acd3ea
                               at /checkout/src/libstd/thread/mod.rs:394
  11:     0x557ec4c259ca - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hbd153f29ff9b4619
                               at /checkout/src/libstd/panic.rs:296
  12:     0x557ec4c95172 - std::panicking::try::do_call::h29276d44934dceb7
                               at /checkout/src/libstd/panicking.rs:480
  13:     0x557ec630301c - __rust_maybe_catch_panic
                               at /checkout/src/libpanic_unwind/lib.rs:99
  14:     0x557ec4c94b6c - std::panicking::try::hca46a7b8aff4134a
                               at /checkout/src/libstd/panicking.rs:459
  15:     0x557ec4c90145 - std::panic::catch_unwind::h416e4480345fc2b2
                               at /checkout/src/libstd/panic.rs:361
  16:     0x557ec4c9363e - std::thread::Builder::spawn::{{closure}}::hc1b6f5e19b9ae0e2
                               at /checkout/src/libstd/thread/mod.rs:393
  17:     0x557ec4d37573 - <F as alloc::boxed::FnBox<A>>::call_box::h4307a547c94c88a6
                               at /checkout/src/liballoc/boxed.rs:682
  18:     0x557ec62faddb - std::sys::imp::thread::Thread::new::thread_start::hf0e7be833820328e
                               at /checkout/src/liballoc/boxed.rs:692
                               at /checkout/src/libstd/sys_common/thread.rs:21
                               at /checkout/src/libstd/sys/unix/thread.rs:84
  19:     0x7f2f1ef926b9 - start_thread
  20:     0x7f2f1eab23dc - clone
  21:                0x0 - <unknown>
```

**Comment by ignopeverell** [comment]
This is around the exact same lines:

```
Dec 17 13:34:42.019 DEBG (Server ID: Port 13414) Built new block with 1 inputs and 3 outputs, difficulty: 565
Dec 17 13:34:42.500 DEBG handle_payload: Block 15f77e94
Dec 17 13:34:42.501 DEBG Received block 15f77e94 at 21328 from network, going to process.
Dec 17 13:34:42.503 DEBG pipe: process_block 15f77e94 at 21328 with 1 inputs and 3 outputs.
Dec 17 13:34:42.504 DEBG Block 15f77e94 at 21328 is unfit at this time: already known
Dec 17 13:34:42.505 DEBG Block 15f77e94 refused by chain: Unfit("already known")
Dec 17 13:34:42.520 DEBG Starting new sumtree extension.
Dec 17 13:34:42.521 DEBG Error returned, discarding sumtree extension.
Dec 17 13:34:42.521 ERRO Error setting sumtree root to build a block: AlreadySpent
thread 'miner' panicked at 'Box<Any>', grin/src/miner.rs:599:4
stack backtrace:
   0:     0x55b7e25269b3 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::hfc7985b08e763a82
                               at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x55b7e25229a4 - std::sys_common::backtrace::_print::h16a1db02a59ead63
                               at /checkout/src/libstd/sys_common/backtrace.rs:71
   2:     0x55b7e2529f23 - std::panicking::default_hook::{{closure}}::h48ecee46f2eefc30
                               at /checkout/src/libstd/sys_common/backtrace.rs:60
                               at /checkout/src/libstd/panicking.rs:381
   3:     0x55b7e2529c92 - std::panicking::default_hook::hb4c92ae8d005ca44
                               at /checkout/src/libstd/panicking.rs:397
   4:     0x55b7e252a3e7 - std::panicking::rust_panic_with_hook::h25d461655d60b1a5
                               at /checkout/src/libstd/panicking.rs:611
   5:     0x55b7e0ec31f9 - std::panicking::begin_panic::hc7980fb899a9ae73
                               at /checkout/src/libstd/panicking.rs:572
   6:     0x55b7e108566a - grin_grin::miner::Miner::build_block::h50ccdac7ded23bfd
                               at grin/src/miner.rs:599
   7:     0x55b7e10827b3 - grin_grin::miner::Miner::run_loop::h2e1d64d84a980999
                               at grin/src/miner.rs:450
   8:     0x55b7e1070ba6 - grin_grin::server::Server::start_miner::{{closure}}::hb649cc1486ca9c15
                               at grin/src/server.rs:226
   9:     0x55b7e0ea712a - std::sys_common::backtrace::__rust_begin_short_backtrace::hb991bd8e33204edc
                               at /checkout/src/libstd/sys_common/backtrace.rs:136
  10:     0x55b7e0ec2da3 - std::thread::Builder::spawn::{{closure}}::{{closure}}::ha855a6e5c2acd3ea
                               at /checkout/src/libstd/thread/mod.rs:394
  11:     0x55b7e0e5477a - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hbd153f29ff9b4619
                               at /checkout/src/libstd/panic.rs:296
  12:     0x55b7e0ec3f22 - std::panicking::try::do_call::h29276d44934dceb7
                               at /checkout/src/libstd/panicking.rs:480
  13:     0x55b7e25314cc - __rust_maybe_catch_panic
                               at /checkout/src/libpanic_unwind/lib.rs:99
  14:     0x55b7e0ec391c - std::panicking::try::hca46a7b8aff4134a
                               at /checkout/src/libstd/panicking.rs:459
  15:     0x55b7e0ebeef5 - std::panic::catch_unwind::h416e4480345fc2b2
                               at /checkout/src/libstd/panic.rs:361
  16:     0x55b7e0ec23ee - std::thread::Builder::spawn::{{closure}}::hc1b6f5e19b9ae0e2
                               at /checkout/src/libstd/thread/mod.rs:393
  17:     0x55b7e0f66323 - <F as alloc::boxed::FnBox<A>>::call_box::h4307a547c94c88a6
                               at /checkout/src/liballoc/boxed.rs:682
  18:     0x55b7e252928b - std::sys::imp::thread::Thread::new::thread_start::hf0e7be833820328e
                               at /checkout/src/liballoc/boxed.rs:692
                               at /checkout/src/libstd/sys_common/thread.rs:21
                               at /checkout/src/libstd/sys/unix/thread.rs:84
  19:     0x7fb10d1c76b9 - start_thread
  20:     0x7fb10cce73dc - clone
  21:                0x0 - <unknown>
```

**Comment by Makadu59** [comment]
here those i get : 
[first](https://pastebin.com/9NiaN30s)
[second](https://pastebin.com/k4Jbt7Ce)
[third](https://pastebin.com/zMHLJHiz)

**Comment by ignopeverell** [comment]
Fixed in abcecd8
